src/urr.cpp

```diff
@@ -0,0 +1,56 @@
+#include "openmc/urr.h"
+
+#include <iostream>
+
+namespace openmc {
+
+void
+UrrData::from_hdf5(hid_t group_id)
+{
+  // Read interpolation and other flags
+  int interp_temp;
+  read_attribute(group_id, "interpolation", interp_temp);
+  switch (interp_temp) {
+  case static_cast<int>(Interpolation::histogram):
+    interp_ = Interpolation::histogram;
+    break;
+  case static_cast<int>(Interpolation::lin_lin):
+    interp_ = Interpolation::lin_lin;
+    break;
+  case static_cast<int>(Interpolation::lin_log):
+    interp_ = Interpolation::lin_log;
+    break;
+  case static_cast<int>(Interpolation::log_lin):
+    interp_ = Interpolation::log_lin;
+    break;
+  case static_cast<int>(Interpolation::log_log):
+    interp_ = Interpolation::log_log;
+  }
+
+  // read the metadata
+  read_attribute(group_id, "inelastic", inelastic_flag_);
+  read_attribute(group_id, "absorption", absorption_flag_);
+  int temp_multiply_smooth;
+  read_attribute(group_id, "multiply_smooth", temp_multiply_smooth);
+  multiply_smooth_ = (temp_multiply_smooth == 1);
+
+  // read the enrgies at which tables exist
+  hid_t dset = open_dataset(group_id, "energy");
+  hsize_t dims[1];
+  get_shape(dset, dims);
+  close_dataset(dset);
+  n_energy_ = static_cast<int>(dims[0]);
+  energy_ = xt::xtensor<double, 1>({dims[0]}, 0.);
+  read_dataset_as_shape(group_id, "energy", energy_);
+
+  // Read URR tables
+  dset = open_dataset(group_id, "table");
+  hsize_t dims3[3];
+  get_shape(dset, dims3);
+  close_dataset(dset);
+  xt::xarray<double> temp_arr({dims3[0], dims3[1], dims3[2]}, 0.);
+  read_dataset(group_id, "table", temp_arr);
+  prob_ = temp_arr;
```

Comment:
Instead of doing it this way, let's just have an overload for `read_dataset` for `xtensor`. There is already one for `xarray` so it will basically look the same. In fact, I was able to get a single overload for `write_dataset` that works for both `xarray` and `xtensor`, so you might be able to generalize the existing overload of `read_dataset` for `xarray`s to just work for both. Anyway, with that change, this would be as simple as:
```C++
read_dataset(group_id, "table", prob_);
```

Reply:
I am having link-time trouble getting that same `write_dataset` approach to work.  Specifically, using `using T = typename D:::value_type`, the linker claims
`lib/libopenmc.so: undefined reference to openmc::H5TypeMap<xt::xtensor_container<xt::uvector<double, std::allocator<double> >, 3ul, (xt::layout_type)1, xt::xtensor_expression_tag> >::type_id`

This tells me that T isn't getting `value_type` but the whole class basically.  

For posterity, this is what I've implemented as `read_dataset`:

```// Template for xarray, xtensor, etc.
template<typename D> inline void
read_dataset(hid_t obj_id, const char* name, const xt::xcontainer<D>& arr,
             bool indep=false)
{
  using T = typename D::value_type;
  read_dataset(obj_id, name, H5TypeMap<T>::type_id, arr.data(), indep);
}
```

I will momentarily push all the other changes.

Reply:
If you can't get a generic `xcontainer` overload to work, then let's just add an overload for `xtensor` that looks like the existing `xarray` overload for now to at least simplify the code here.

Reply:
Ok, I have an xtensor overload in there now